### PR TITLE
fuzz-systemctl-parse-argv: avoid "leak" of bus object

### PR DESCRIPTION
**Pull request by keszybz**
Memory sanitizer would report leaked memory from --boot-load-entry=help.

Maybe we should disable all bus connections from the fuzzer? It seems not
appropriate to communicate with logind. OTOH, in a real fuzzing environment
this call should just fail, so maybe that's OK.

I think this should help with the fuzzing failures reported in CI.

### CODE REVIEW

**Comment by keszybz** [comment]
I found another leak.

**Comment by poettering** [comment]
this should probably also call static_destruct() and mac_selinux_finish().

**Comment by poettering** [comment]
i.e. everything that systemctl.c's main() function calls at the end plus the stuff from _DEFINE_MAIN_FUNCTION() that undoes the stuff the cmdline parsers set up

**Comment by keszybz** [comment]
> this should probably also call static_destruct() and mac_selinux_finish().

Easier said that done ;) I've been working on this for a few hours… But a patch is coming up.

**Comment by keszybz** [comment]
> this should probably also call static_destruct() and mac_selinux_finish().

Updated. The diffstat is a bit bigger now… It essentially implements the call to `static_destruct()`. I also did some unrelated cleanup in code that I was touching anyway.

File: src/analyze/analyze.c [hunk 1]
```diff
@@ -92,7 +92,7 @@ static usec_t arg_base_time = USEC_INFINITY;
 STATIC_DESTRUCTOR_REGISTER(arg_dot_from_patterns, strv_freep);
 STATIC_DESTRUCTOR_REGISTER(arg_dot_to_patterns, strv_freep);
 
-struct boot_times {
+typedef struct BootTimes {
```
Comment (poettering) [inline src/analyze/analyze.c:95]: once upon a time we had this rule that noone followed that structs specific to a source file should use "struct" and no "typedef" and be lower_case and stuff exported to other .c files should use typedefs and user CamelCase. But i guess it's a mess right now, and the rule might not be the smartest rule to have, hence looks ok to me

File: src/resolve/resolved-dns-transaction.c [hunk 2]
```diff
@@ -141,23 +141,23 @@ DnsTransaction* dns_transaction_free(DnsTransaction *t) {
...
-            set_isempty(t->notify_transactions_done)) {
-                dns_transaction_free(t);
-                return false;
-        }
+            set_isempty(t->notify_transactions_done))
+                return dns_transaction_free(t);
 
-        return true;
+        return t;
 }
```
Comment (poettering) [inline src/resolve/resolved-dns-transaction.c:161]: interesting model

File: src/network/networkd-util.c [hunk 3]
```diff
@@ -155,10 +155,6 @@ int network_config_section_new(const char *filename, unsigned line, NetworkConfi
         return 0;
 }
 
-void network_config_section_free(NetworkConfigSection *cs) {
-        free(cs);
-}
```
Comment (poettering) [inline src/network/networkd-util.c:160]: i wonder if we have similar cases elsewhere…

File: src/analyze/analyze.c [hunk 4]
```diff
@@ -213,10 +213,10 @@ static int compare_unit_start(const UnitTimes *a, const UnitTimes *b) {
         return CMP(a->activating, b->activating);
 }
 
-static void unit_times_free(UnitTimes *t) {
-        for (UnitTimes *p = t; p->has_data; p++)
+static UnitTimes* unit_times_free(UnitTimes *t) {
```
Comment (poettering) [inline src/analyze/analyze.c]: maybe add a comment here, that this destroy the whole frickin' array. The function is called so innocently, one might not expect that

File: src/basic/strbuf.c [hunk 5]
```diff
@@ -65,13 +65,13 @@ void strbuf_complete(struct strbuf *str) {
 }
 
 /* clean up everything */
-void strbuf_cleanup(struct strbuf *str) {
+struct strbuf* strbuf_cleanup(struct strbuf *str) {
```
Comment (poettering) [inline src/basic/strbuf.c]: maybe rename strbuf_free?

File: src/busctl/busctl.c [hunk 6]
```diff
@@ -780,24 +780,22 @@ static int member_compare_funcp(Member * const *a, Member * const *b) {
...
-
 DEFINE_TRIVIAL_CLEANUP_FUNC(Member*, member_free);
 
-static void member_set_free(Set *s) {
-        set_free_with_destructor(s, member_free);
+static Set* member_set_free(Set *s) {
+        return set_free_with_destructor(s, member_free);
 }
-
 DEFINE_TRIVIAL_CLEANUP_FUNC(Set*, member_set_free);
```
Comment (poettering) [inline src/busctl/busctl.c:799]: i guess this should use a destructor in its hash_ops...

File: src/basic/macro.h [hunk 7]
```diff
@@ -405,10 +405,20 @@ static inline int __coverity_check_and_return__(int condition) {
                 func(p);                                        \
         }
 
+/* When func() returns the appropriate type */
```
Comment (poettering) [inline src/basic/macro.h]: comment should say *NULL cast to the appropriate type*.

i.e. the NULL matters

File: src/systemctl/systemctl-compat-shutdown.c [hunk 8]
```diff
@@ -132,9 +132,10 @@ int shutdown_parse_argv(int argc, char *argv[]) {
                 wall = argv + optind + 1;
 
         if (wall) {
-                arg_wall = strv_copy(wall);
-                if (!arg_wall)
+                char **copy = strv_copy(wall);
+                if (!copy)
                         return log_oom();
+                strv_free_and_replace(arg_wall, copy);
```
Comment (poettering) [inline src/systemctl/systemctl-compat-shutdown.c:138]: maybe add a helper strv_free_and_copy(), taking inspiration from free_and_strdup()?

**Comment by poettering** [comment]
lgtm, just some nitpicks

File: src/analyze/analyze.c (hunk 1 above)
Comment (keszybz) [reply src/analyze/analyze.c:95]: Oh, OK. Anyway, I don't think this is even remotely true. We should probably convert the remaining few instances without typedef too.

File: src/basic/macro.h (hunk 7 above)
Comment (keszybz) [reply src/basic/macro.h]: Actually it's not NULL — in some cases it's e.g. -1. If it was always NULL, I wouldn't need the second variant of DEFINE_TRIVIAL_CLEANUP_FUNC. I'll try to improve the comment.

File: src/analyze/analyze.c [hunk 9]
```diff
@@ -414,28 +411,28 @@ static int acquire_time_data(sd_bus *bus, struct unit_times **out) {
...
-                { "Architecture",   "s", NULL, offsetof(struct host_info, architecture)   },
+                { "Virtualization", "s", NULL, offsetof(HostInfo, virtualization) },
+                { "Architecture",   "s", NULL, offsetof(HostInfo, architecture)   },
                 {}
         };
 
         _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
         _cleanup_(sd_bus_flush_close_unrefp) sd_bus *system_bus = NULL;
-        _cleanup_(free_host_infop) struct host_info *host;
+        _cleanup_(free_host_infop) HostInfo *host;
```
Comment (yuwata) [inline src/analyze/analyze.c]: I'd like to add ` = NULL` here for `host`. Of course, it is redundant now. But one day, some code may be added before `host = new0()`.

**Review by yuwata — APPROVED** [review]
LGTM. One minor comment.

**Comment by yuwata** [comment]
CI fails.
```
In file included from ../src/basic/fd-util.h:9,
                 from ../src/core/apparmor-setup.c:11:
../src/core/apparmor-setup.c: In function ‘aa_policy_cache_unrefp’:
../src/basic/macro.h:412:28: error: void value not ignored as it ought to be
  412 |                         *p = func(*p);                          \
      |                            ^
../src/core/apparmor-setup.c:19:1: note: in expansion of macro ‘DEFINE_TRIVIAL_CLEANUP_FUNC’
   19 | DEFINE_TRIVIAL_CLEANUP_FUNC(aa_policy_cache *, aa_policy_cache_unref);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/core/apparmor-setup.c: In function ‘aa_features_unrefp’:
../src/basic/macro.h:412:28: error: void value not ignored as it ought to be
  412 |                         *p = func(*p);                          \
      |                            ^
../src/core/apparmor-setup.c:20:1: note: in expansion of macro ‘DEFINE_TRIVIAL_CLEANUP_FUNC’
   20 | DEFINE_TRIVIAL_CLEANUP_FUNC(aa_features *, aa_features_unref);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: src/network/networkd-util.c (hunk 3 above)
Comment (keszybz) [reply src/network/networkd-util.c:160]: Not many I think. Anyway, the compiler should mostly optimize this away when doing lto. So it's mostly a matter of saving a few lines of sources.

**Comment by keszybz** [comment]
Updated:
- two commits to rename the functions that @poettering mentioned
- I fixed all the "hidden" comments (NULL initialization, apparmor failure, strbuf_cleanup rename, unit_times_freep rename)

File: src/busctl/busctl.c (hunk 6 above)
Comment (keszybz) [reply src/busctl/busctl.c:799]: That's require checking what the code expects... I'm leaving that for later.

File: src/systemctl/systemctl-compat-shutdown.c (hunk 8 above)
Comment (keszybz) [reply src/systemctl/systemctl-compat-shutdown.c:138]: I'm leaving that for later... Such cleanups can always be done, but the goal here was to make the fuzzer not fail to quickly ;)

**Comment by yuwata** [comment]
LGTM.

**Pull request merged**